### ISSUE TITLE
TS 5.8: False positives for conditional expression in return type check

### ISSUE THREAD

**Issue by AlCalzone** [microsoft/TypeScript#61337]
### 🔎 Search Terms

5.8, return, type

### 🕗 Version & Regression Information

- This changed between versions 5.7.3 and 5.8.2

### 💻 Code

```ts
export type ReturnTypeOrStatic<T> = T extends (...args: any[]) => infer R ? R
	: T;

export function evalOrStatic<T>(
	fnOrConst: T,
	...args: any[]
): ReturnTypeOrStatic<T> {
	return typeof fnOrConst === "function" ? fnOrConst(...args) : fnOrConst;
}
```

### 🙁 Actual behavior

Error under `fnOrConst` at the end of the conditional: 
```
Type 'T' is not assignable to type 'ReturnTypeOrStatic<T>'
```

### 🙂 Expected behavior

no error. If `fnOrConst` is a function, its return value (type `R`) is returned, otherwise it is returned directly (type `T`).

**Comment by Andarist** [comment]
>  If fnOrConst is a function, its return value (type R) is returned, otherwise it is returned directly (type T).

This is true *but* TS doesn't have a concept of negated types today so it can't quite reason about the "else" branch of this. 

This is an expected breaking change. It was announced [here](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-5-8.html#granular-checks-for-branches-in-return-expressions). Note that your code has never worked in a similar situation (that logical-wise is actually the same):
```ts
export type ReturnTypeOrStatic<T> = T extends (...args: any[]) => infer R
  ? R
  : T;

export function evalOrStatic2<T>(
  fnOrConst: T,
  ...args: any[]
): ReturnTypeOrStatic<T> {
  if (typeof fnOrConst === "function") {
    return fnOrConst(...args);
  }
  return fnOrConst; // errors
}
```

You can see it on the playground [here](https://www.typescriptlang.org/play/?ts=5.7.0-dev.20241105#code/KYDwDg9gTgLgBDAnmYcBKwYFcoDsAqywA8lAMowCGMAlgMYA8+AfHALxz5ygzC4AmAZzgAKAHQTKUAOaCAXHEq5EAbQC6ASnasauAGbAo6AFBw4AfhNmF+ANzHjoSLDh6suOrQi5uAN0oANqQU1PRMzCKmrrikAMLegjA2ADRREmJSsgpKqmrGGgoY2HiEKMFUtIwscADeUVCYOD5IKBB60XEJ8Gw9cABEbh5euH0WHVDxuInikjKCWgp6MRNd9gC+Dk7Q8IOeNN5+geWhdABM4ZFmS51TSZypZumZ8orK6vmFjSVEx5XhtVEaO0RC1gG1xpNEuxegN3HtvH0tHUzGYGsUfNcVrcZhk5hp7GYNqivhjlpCYOtjEA)
